## Connecting to SQL database and validating tables

In [521]:
# Connect to SQL database 

import sqlalchemy
from sqlalchemy import create_engine
%load_ext sql
%sql postgresql://postgres:password@localhost/hawaii_airbnb
engine = sqlalchemy.create_engine('postgresql://postgres:password@localhost:5432/hawaii_airbnb')

listings.to_sql('listings', con=engine, if_exists='replace', index=False)
reviews.to_sql('reviews', con=engine, if_exists='replace', index=False)
hosts.to_sql('hosts', con=engine, if_exists='replace', index=False)

# Check that tables are properly loaded into database hawaii_airbnb
%sql SELECT table_name FROM information_schema.tables WHERE table_schema = 'public';

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
 * postgresql://postgres:***@localhost/hawaii_airbnb
3 rows affected.


[('listings',), ('reviews',), ('hosts',)]

In [522]:
%%sql

-- Check to make sure data types were captured correctly in SQL tables
SELECT table_name, column_name, data_type
FROM information_schema.columns
WHERE table_name IN ('listings', 'hosts', 'reviews')
ORDER BY table_name, data_type;

 * postgresql://postgres:***@localhost/hawaii_airbnb
63 rows affected.


[('hosts', 'host_acceptance_rate', 'bigint'),
 ('hosts', 'host_id', 'bigint'),
 ('hosts', 'host_response_rate', 'bigint'),
 ('hosts', 'host_is_superhost', 'boolean'),
 ('hosts', 'host_listings_count', 'double precision'),
 ('hosts', 'host_verifications', 'text'),
 ('hosts', 'host_response_time', 'text'),
 ('hosts', 'host_identity_verified', 'text'),
 ('hosts', 'host_about', 'text'),
 ('hosts', 'host_location', 'text'),
 ('hosts', 'host_neighbourhood', 'text'),
 ('hosts', 'host_name', 'text'),
 ('hosts', 'host_has_profile_pic', 'text'),
 ('hosts', 'host_since', 'timestamp without time zone'),
 ('listings', 'availability_60', 'bigint'),
 ('listings', 'availability_365', 'bigint'),
 ('listings', 'number_of_reviews', 'bigint'),
 ('listings', 'number_of_reviews_ltm', 'bigint'),
 ('listings', 'number_of_reviews_l30d', 'bigint'),
 ('listings', 'calculated_host_listings_count', 'bigint'),
 ('listings', 'bathroom_shared', 'bigint'),
 ('listings', 'availability_30', 'bigint'),
 ('listings', 'availability_90', 'bigint'),
 ('listings', 'host_id', 'bigint'),
 ('listings', 'accommodates', 'bigint'),
 ('listings', 'calculated_host_listings_count_entire_homes', 'bigint'),
 ('listings', 'calculated_host_listings_count_private_rooms', 'bigint'),
 ('listings', 'calculated_host_listings_count_shared_rooms', 'bigint'),
 ('listings', 'minimum_nights', 'bigint'),
 ('listings', 'maximum_nights', 'bigint'),
 ('listings', 'instant_bookable', 'boolean'),
 ('listings', 'has_availability', 'boolean'),
 ('listings', 'bedrooms', 'double precision'),
 ('listings', 'review_scores_checkin', 'double precision'),
 ('listings', 'review_scores_communication', 'double precision'),
 ('listings', 'review_scores_location', 'double precision'),
 ('listings', 'review_scores_value', 'double precision'),
 ('listings', 'reviews_per_month', 'double precision'),
 ('listings', 'bathroom_num', 'double precision'),
 ('listings', 'id', 'double precision'),
 ('listings', 'beds', 'double precision'),
 ('listings', 'price', 'double precision'),
 ('listings', 'review_scores_rating', 'double precision'),
 ('listings', 'review_scores_accuracy', 'double precision'),
 ('listings', 'review_scores_cleanliness', 'double precision'),
 ('listings', 'name', 'text'),
 ('listings', 'neighbourhood_cleansed', 'text'),
 ('listings', 'island', 'text'),
 ('listings', 'property_type', 'text'),
 ('listings', 'room_type', 'text'),
 ('listings', 'neighborhood_overview', 'text'),
 ('listings', 'amenities', 'text'),
 ('listings', 'description', 'text'),
 ('listings', 'bathrooms_text', 'text'),
 ('listings', 'last_review', 'timestamp without time zone'),
 ('listings', 'first_review', 'timestamp without time zone'),
 ('reviews', 'reviewer_id', 'bigint'),
 ('reviews', 'listing_id', 'double precision'),
 ('reviews', 'id', 'double precision'),
 ('reviews', 'comments', 'text'),
 ('reviews', 'reviewer_name', 'text'),
 ('reviews', 'date', 'text'),
 ('reviews', 'language', 'text')]

## Table features and descriptions

### Listings
* id - Airbnb's unique identifier for the listing
* name - Name of the listing
* description - Detailed description of the listing
* neighborhood_overview - Host's description of the neighbourhood
* host_id - Airbnb's unique identifier for the host/user
* neighbourhood_cleansed - The neighbourhood as geocoded using the latitude and longitude against neighborhoods as defined by open or public digital shapefiles
* island - Island of Hawaii where the listing is located 
* property_type - Host-selected property type
* room_type - Accommodation type, selected from three options (Entire home/apt, Prive room, Shared room)
* accommodates - The maximum capacity of the listing
* bathrooms_text - Number of bathrooms (text)
* bathroom_num - Number of bathrooms (numerical)
* bathroom_shared - Whether bathroom is shared (1) or not (0)
* bedrooms - Number of bedrooms
* beds - Number of beds
* amenities - Amenities available
* price - Daily price in local currency
* minimum_nights - minimum number of night stay for the listing
* maximum_nights - maximum number of night stay for the listing
* has_availability - whether the listing had availability as of the day of scrape 
* availability_30 - The availability of the listing 30 days in the future as determined by the calendar
* availability_60 - The availability of the listing 60 days in the future as determined by the calendar
* availability_90 - The availability of the listing 90 days in the future as determined by the calendar
* availability_365 - The availability of the listing 365 days in the future as determined by the calendar
* number_of_reviews - The number of reviews the listing has
* number_of_reviews_ltm - The number of reviews the listing has (in the last 12 months)
* number_of_reviews_l30d - The number of reviews the listing has (in the last 30 days)
* first_review - The date of the first/oldest review
* last_review - The date of the last/newest review
* review_scores_rating - Average overall rating on listing
* review_scores_accuracy - Average accuracy rating on listing
* review_scores_cleanliness - Average cleanliness rating on listing
* review_scores_checkin - Average check-in rating on listing
* review_scores_communication - Average host communication rating on listing
* review_scores_location - Average location rating on listing
* review_scores_value - Average value rating on listing
* instant_bookable - Whether the guest can automatically book the listing without the host needing to accept their booking request, an indicator of a commercial listing
* calculated_host_listings_count - The number of listings the host has in the current scrape, in the city/region geography
* calculated_host_listings_count_entire_homes - The number of Entire home/apt listings the host has in the current scrape, in the city/region geography
* calculated_host_listings_count_private_rooms - The number of Private room listings the host has in the current scrape, in the city/region geography
* calculated_host_listings_count_shared_rooms - The number of Shared room listings the host has in the current scrape, in the city/region geography
* reviews_per_month - The number of reviews the listing has over the lifetime of the listing

### Reviews
* listing_id - Airbnb's unique identifier for the listing
* id - Unique idenitifier for review
* date - Date of review
* reviewer_id - Reviewer's unique identifier
* reviewer_name - Name of review
* comments - Review left by reviewer

### Hosts
* host_id - Airbnb's unique identifier for the host/user
* host_name - First name of the host
* host_since - The date the host/user was created, for hosts that are Airbnb guests this could be the date they registered as a guest
* host_location - The host's self reported location
* host_about - Description about the host
* host_response_time - Average host response time to guest inquiries
* host_response_rate - Average host response rate to guest inquiries
* host_acceptance_rate - That rate at which a host accepts booking requests
* host_is_superhost - Whether host is a superhost or not, meaning they were selected by Airbnb as a host that is extremely welcoming and provides extraordinary experiences for their guests
* host_neighbourhood - Neighborhood where host is located
* host_listings_count - The number of listings the host has (per Airbnb calculations)
* host_verifications - How the host has been verified 
* host_has_profile_pic - Whether host has profile picture or not
* host_identity_verified - Whether host has had identity verified 

## Exploring data 

I will take a preliminary look at the tables in the database by using queries to answer simple questions such as "How many listings are there on each island?" and "What is the most common number of bedrooms and bathrooms per listing?". 

### How many listings are there on each island, and what percentage of total listings does each island contain?

In [523]:
%%sql

SELECT island, COUNT(id), ROUND(COUNT(id) * 100.0 / (SELECT COUNT(*) FROM listings), 2) as percent_of_total
FROM listings
GROUP BY island
ORDER BY percent_of_total DESC;

 * postgresql://postgres:***@localhost/hawaii_airbnb
4 rows affected.


[('Maui', 9513, Decimal('31.64')),
 ('Oahu', 8867, Decimal('29.49')),
 ('Hawaii', 7045, Decimal('23.43')),
 ('Kauai', 4639, Decimal('15.43'))]

Maui has the most listings with 31.64% of the total, and Kauai has the least listings with 15.43% of the total.

### How many hosts have listings in Hawaii by island?

In [524]:
%%sql

SELECT island, COUNT (DISTINCT host_id) AS unique_hosts
FROM listings
GROUP BY island
ORDER BY unique_hosts DESC;

 * postgresql://postgres:***@localhost/hawaii_airbnb
4 rows affected.


[('Oahu', 2744), ('Hawaii', 2678), ('Maui', 2236), ('Kauai', 1020)]

Oahu has the most Airbnb hosts, while Kauai has the least. This tells us that hosts in Maui are more likely to have multiple listings compared to hosts in Oahu, as Maui has more listings overall but less hosts.

### What is the most common property type, room type, accommodation capacity, number of bedrooms and number of bathrooms for listings on any of the islands?

In [525]:
%%sql

SELECT feature, most_popular
FROM (
  SELECT 'property_type' AS feature, property_type AS most_popular,
         RANK() OVER (ORDER BY COUNT(property_type) DESC) AS rank
  FROM listings
  GROUP BY property_type
  UNION
  SELECT 'room_type' AS feature, room_type AS most_popular,
         RANK() OVER (ORDER BY COUNT(room_type) DESC) AS rank
  FROM listings
  GROUP BY room_type
  UNION
  SELECT 'accommodates' AS feature, accommodates::text AS most_popular,
         RANK() OVER (ORDER BY COUNT(accommodates) DESC) AS rank
  FROM listings
  GROUP BY accommodates
  UNION
  SELECT 'bedrooms' AS feature, bedrooms::text AS most_popular,
         RANK() OVER (ORDER BY COUNT(bedrooms) DESC) AS rank
  FROM listings
  GROUP BY bedrooms
  UNION
  SELECT 'bathroom_num' AS feature, bathroom_num::text AS most_popular,
         RANK() OVER (ORDER BY COUNT(bathroom_num) DESC) AS rank
  FROM listings
  GROUP BY bathroom_num
) AS subquery
WHERE rank = 1
ORDER BY feature;


 * postgresql://postgres:***@localhost/hawaii_airbnb
5 rows affected.


[('accommodates', '4'),
 ('bathroom_num', '1'),
 ('bedrooms', '1'),
 ('property_type', 'Entire condo'),
 ('room_type', 'Entire home/apt')]

Listings most commonly accommodate 4 people, have 1 bathroom and 1 bedroom, are of the property type "Entire condo", and are of the room type "Entire home/apt". 

Now that we have described the features in the tables and answered some simple questions using SQL queries, I will continue with a more in-depth analysis of the data.